File: py/vec3.py
import math

class Vec3:
    def __init__(self, comp):
        self.comp = tuple(comp)

    def __repr__(self) -> str:
        return str(self.comp)

    def x(self):
        return self.comp[0]

    def y(self):
        return self.comp[1]
        
    def z(self):
        return self.comp[2]
        
    def verif_type(other, op_name):
        if type(other) is not Vec3:
            raise NotImplementedError(
                'cannot {} Vec3 and {}'.format(op_name, type(other)))

    def __eq__(self, other):
        Vec3.verif_type(other, 'eq')
        return self.comp == other.comp

    def __add__(self, other):
        Vec3.verif_type(other, 'add')
        return Vec3(v + other.comp[i] for i, v in enumerate(self.comp))

    def __neg__(self):
        return Vec3(-v for v in self.comp)

    def __sub__(self, other):
        Vec3.verif_type(other, 'sub')
        return Vec3(v - other.comp[i] for i, v in enumerate(self.comp))

    def __mul__(self, other):
        if type(other) is not int and type(other) is not float:
            raise NotImplementedError(
                'cannot mul Vec3 and {}'.format(type(other)))
        return Vec3(v * other for v in self.comp)

    def __rmul__(self, other):
        return self.__mul__(other)

    def __truediv__(self, other):
        if type(other) is not int and type(other) is not float:
            raise NotImplementedError(
                'cannot div Vec3 and {}'.format(type(other)))
        return Vec3(v / other for v in self.comp)

    def dot(self, other):
        Vec3.verif_type(other, 'dot')
        return sum(v * other.comp[i] for i, v in enumerate(self.comp))

    def cross(self, other):
        Vec3.verif_type(other, 'cross')
        ax, ay, az = self.comp
        bx, by, bz = other.comp
        return Vec3((ay * bz - az * by, az * bx - ax * bz, ax * by - ay * bx))

    def norm(self):
        return math.sqrt(self.dot(self))

    def unity(self):
        return self / self.norm()

    def compare(self, other, tol):
        Vec3.verif_type(other, 'compare')
        return (self - other).norm() < abs(tol)

    def at_len(self, t):
        if type(t) is not int and type(t) is not float:
            raise NotImplementedError(
                'cannot at_len with {}'.format(type(t)))
        return self.unity() * t

    def abs(self):
        return Vec3(abs(x) for x in self.comp)

File: rust/src/vec3.rs
use std::cmp::PartialEq;
use std::ops;

#[derive(Debug, Clone, Copy)]
pub struct Vec3 {
    x: f32,
    y: f32,
    z: f32,
}

impl Vec3 {
    pub fn new(x: f32, y: f32, z: f32) -> Self {
        Vec3 { x, y, z }
    }

    pub fn x(self) -> f32 {
        self.x
    }

    pub fn y(self) -> f32 {
        self.y
    }

    pub fn z(self) -> f32 {
        self.z
    }

    pub fn dot(self, other: Vec3) -> f32 {
        self.x * other.x + self.y * other.y + self.z * other.z
    }

    pub fn cross(self, other: Vec3) -> Vec3 {
        let a = self;
        let b = other;
        Vec3 {
            x: a.y * b.z - a.z * b.y,
            y: a.z * b.x - a.x * b.z,
            z: a.x * b.y - a.y * b.x,
        }
    }

    pub fn norm(&self) -> f32 {
        (self.x * self.x + self.y * self.y + self.z * self.z).sqrt()
    }

    pub fn compare(self, other: Vec3, tol: f32) -> bool {
        (self - other).norm() < tol.abs()
    }

    pub fn unity(&self) -> Vec3 {
        *self / self.norm()
    }

    pub fn at_len(self, t: f32) -> Vec3 {
        self.unity() * t
    }
}

impl PartialEq for Vec3 {
    fn eq(&self, other: &Self) -> bool {
        self.x == other.x && self.y == other.y && self.z == other.z
    }
}

impl ops::Add for Vec3 {
    type Output = Self;

    fn add(self, other: Self) -> Self {
        Self {
            x: self.x + other.x,
            y: self.y + other.y,
            z: self.z + other.z,
        }
    }
}

impl ops::Sub for Vec3 {
    type Output = Self;

    fn sub(self, other: Self) -> Self {
        Self {
            x: self.x - other.x,
            y: self.y - other.y,
            z: self.z - other.z,
        }
    }
}

impl ops::Neg for Vec3 {
    type Output = Self;

    fn neg(self) -> Self {
        Self {
            x: -self.x,
            y: -self.y,
            z: -self.z,
        }
    }
}

impl ops::Mul<f32> for Vec3 {
    type Output = Self;

    fn mul(self, other: f32) -> Self {
        Self {
            x: other * self.x,
            y: other * self.y,
            z: other * self.z,
        }
    }
}

impl ops::Mul<Vec3> for f32 {
    type Output = Vec3;

    fn mul(self, other: Vec3) -> Vec3 {
        Vec3 {
            x: self * other.x,
            y: self * other.y,
            z: self * other.z,
        }
    }
}

impl ops::Div<f32> for Vec3 {
    type Output = Self;

    fn div(self, other: f32) -> Self {
        Self {
            x: self.x / other,
            y: self.y / other,
            z: self.z / other,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_eq() {
        assert_eq!(Vec3::new(1., 2., 3.), Vec3::new(1., 2., 3.));
    }

    #[test]
    fn test_add() {
        assert_eq!(
            Vec3::new(1., 2., 3.) + Vec3::new(4., 5., 6.),
            Vec3::new(5., 7., 9.)
        );
    }

    #[test]
    fn test_sub() {
        assert_eq!(
            Vec3::new(5., 7., 9.) - Vec3::new(4., 5., 6.),
            Vec3::new(1., 2., 3.)
        );
    }

    #[test]
    fn test_neg() {
        assert_eq!(-Vec3::new(1., 2., 3.), Vec3::new(-1., -2., -3.));
    }

    #[test]
    fn test_mul() {
        assert_eq!(Vec3::new(1., 2., 3.) * 2., Vec3::new(2., 4., 6.));
        assert_eq!(2. * Vec3::new(1., 2., 3.), Vec3::new(2., 4., 6.));
    }

    #[test]
    fn test_div() {
        assert_eq!(Vec3::new(2., 4., 6.) / 2., Vec3::new(1., 2., 3.));
    }

    #[test]
    fn test_dot() {
        assert_eq!(Vec3::new(1., 2., 3.).dot(Vec3::new(7., 6., 5.)), 34.);
    }

    #[test]
    fn test_cross() {
        assert_eq!(
            Vec3::new(1., 0., 0.).cross(Vec3::new(0., 1., 0.)),
            Vec3::new(0., 0., 1.)
        );
    }

    #[test]
    fn test_norm() {
        assert_eq!(Vec3::new(9., 12., 112.).norm(), 113.);
    }

    #[test]
    fn test_unity() {
        assert!(
            Vec3::new(3., 4., 5.).unity().compare(
                Vec3::new(0.424, 0.566, 0.707),
                0.001));
    }
}
